sdk/servicebus/service-bus/src/connectionContext.ts

```diff
@@ -180,11 +163,66 @@ export namespace ConnectionContext {
       }
     };
 
-    // Add listeners on the connection object.
-    connectionContext.connection.on(ConnectionEvents.connectionOpen, onConnectionOpen);
-    connectionContext.connection.on(ConnectionEvents.disconnected, disconnected);
-    connectionContext.connection.on(ConnectionEvents.protocolError, protocolError);
-    connectionContext.connection.on(ConnectionEvents.error, error);
+    async function reconnect(connectionContext: ConnectionContext, error?: Error | AmqpError) {
+      const originalConnectionId = connectionContext.connectionId;
+      try {
+        await cleanConnectionContext(connectionContext);
+      } catch {
+        log.error(
+          `[${connectionContext.connectionId}] There was an error closing the connection before reconnecting.`
+        );
+      }
+      // Create a new connection, id, locks, and cbs client.
+      connectionContext.refreshConnection();
+      addConnectionListeners(connectionContext.connection);
+      log.error(
+        `The connection "${originalConnectionId}" has been updated to "${connectionContext.connectionId}".`
+      );
+
+      await delay(Constants.connectionReconnectDelay);
+      // reconnect clients if any
+      for (const id of Object.keys(connectionContext.clientContexts)) {
+        const clientContext = connectionContext.clientContexts[id];
+        log.error(
+          "[%s] calling detached on client '%s'.",
+          connectionContext.connection.id,
+          clientContext.clientId
+        );
+        clientContext.onDetached(error).catch((err) => {
+          log.error(
+            "[%s] An error occurred while reconnecting the sender '%s': %O.",
+            connectionContext.connection.id,
+            clientContext.clientId,
+            err
+          );
+        });
+      }
+    }
+
+    function addConnectionListeners(connection: Connection) {
+      // Add listeners on the connection object.
+      connection.on(ConnectionEvents.connectionOpen, onConnectionOpen);
+      connection.on(ConnectionEvents.disconnected, disconnected);
+      connection.on(ConnectionEvents.protocolError, protocolError);
+      connection.on(ConnectionEvents.error, error);
+    }
+
+    async function cleanConnectionContext(connectionContext: ConnectionContext) {
+      // Remove listeners from the connection object.
+      connectionContext.connection.removeListener(
+        ConnectionEvents.connectionOpen,
+        onConnectionOpen
+      );
+      connectionContext.connection.removeListener(ConnectionEvents.disconnected, disconnected);
+      connectionContext.connection.removeListener(ConnectionEvents.protocolError, protocolError);
+      connectionContext.connection.removeListener(ConnectionEvents.error, error);
+      // Close the cbs session.
+      await connectionContext.cbsSession.close();
+      // Close the connection
+      await connectionContext.connection.close();
```

Comment:
If we are calling `close()` on the connection, then do we need to remove the listeners explicitly? I would expect rhea to do the clean up on the listeners

Reply:
I suspect not, but I can verify that.

Reply:
Turns out we do need to cleanup the listeners, because the connection we're interacting with is a `rhea-promise` connection, and it only removes listeners that it adds itself when you call close on it.